src/nni_manager/training_service/pai_base/paiBaseTrainingService.ts

```diff
@@ -0,0 +1,297 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+'use strict';
+
+import * as fs from 'fs';
+import * as path from 'path';
+import * as request from 'request';
+import * as component from '../../common/component';
+
+import { EventEmitter } from 'events';
+import { Deferred } from 'ts-deferred';
+import { String } from 'typescript-string-operations';
+import { getExperimentId } from '../../common/experimentStartupInfo';
+import { getLogger, Logger } from '../../common/log';
+import {
+    HyperParameters, NNIManagerIpConfig, TrainingService,
+    TrialJobApplicationForm, TrialJobDetail, TrialJobMetric
+} from '../../common/trainingService';
+import { delay, generateParamFileName,
+    getExperimentRootDir, getIPV4Address, getVersion, uniqueString, unixPathJoin } from '../../common/utils';
+import { CONTAINER_INSTALL_NNI_SHELL_FORMAT } from '../common/containerJobData';
+import { TrialConfigMetadataKey } from '../common/trialConfigMetadataKey';
+import { execMkdir, validateCodeDir } from '../common/util';
+import { PAIBaseJobInfoCollector } from './paiBaseJobInfoCollector';
+import { PAIBaseJobRestServer, ParameterFileMeta } from './paiBaseJobRestServer';
+import { PAIBaseClusterConfig, PAIBaseTrialJobDetail } from './paiBaseConfig';
+
+/**
+ * Training Service implementation for OpenPAI (Open Platform for AI)
+ * Refer https://github.com/Microsoft/pai for more info about OpenPAI
+ */
+@component.Singleton
+class PAIBaseTrainingService implements TrainingService {
```

Comment:
Suggest to change this base class to abstract class.

Reply:
fixed.